ztest: allow for user-defined main function

fouge opened the pull request:
The goal is to be able to define and run test suites from anywhere and not obligatorily from test applications.

`main()` function isn't redefined.
User has to call `test_main()` to run the tests.

fouge committed:
ztest: execute tests manually from anywhere

KConfig option is available to allow for explicit execution
of the tests from any application, not only test application
with redefined main function.
To run tests, user has to call ztest_main().

Signed-off-by: Cyril Fougeray <cyril.fougeray@worldcoin.org>

yperess reviewed (changes requested):
This PR doesn't include documentation under doc/ or sample/tests for the new code. Please add those and request review again.

fouge committed:
ztest: test for standalone usage

Added a test case to verify correct execution of ztest
 when used with `CONFIG_ZTEST_STANDALONE_EXECUTION`.
 `ztest_main` has to be called manually, here from the main function.

Signed-off-by: Cyril Fougeray <cyril.fougeray@worldcoin.org>

fouge committed:
ztest: documentation for standalone usage

Documentation to explain how to execute tests from any application.

Signed-off-by: Cyril Fougeray <cyril.fougeray@worldcoin.org>

asemjonovs commented on subsys/testsuite/ztest/src/ztest_new.c:
```diff
@@ -1012,11 +1017,16 @@ void ztest_run_all(const void *state) { ztest_api.run_all(state); }
 
+#ifndef CONFIG_ZTEST_STANDALONE_EXECUTION
 #ifndef KERNEL
 int main(void)
```
I think it may be a bit cleaner if we rename the current `main` functions to `ztest_main`  (or some better name) and create new `main` functions that call the renamed `ztest_main()`.  So in your standalone app, you can invoke `ztest_main` which does everything that the current `main` does.  This way we don't need to sprinkle around the inits and report functions throughout the file.

So something like
```
int ztest_main(void) 
{
	z_init_mock();
	test_main();
	end_report();
	return test_status;
};

#ifndef CONFIG_ZTEST_STANDALONE_EXECUTION
int main(void) 
{ 
  return ztest_main(); 
};
#endif 
```

Would this accomplish the behavior you're looking for?

aaronemassey replied:
+1

fouge replied:
Good idea, I'll do this

fouge replied:
I just pushed new commits
Let me know what you think

fouge commented:
Should I add this feature only with the new API? 🤔

asemjonovs commented on subsys/testsuite/ztest/Kconfig:
```diff
@@ -98,12 +99,21 @@ config ZTEST_ASSERT_HOOK
+config ZTEST_STANDALONE_EXECUTION
+	bool "Execute tests manually from anywhere"
+	help
+	  Run tests from anywhere by calling test_main().
```
Nit: Fix function name

asemjonovs commented on subsys/testsuite/ztest/src/ztest.c:
```diff
@@ -422,6 +422,15 @@ static int run_test(struct unit_test *test)
 
 #endif /* !KERNEL */
 
+void end_report(void)
```
Do we still need to move this?

asemjonovs commented on tests/ztest/base/src/main.c:
```diff
@@ -166,3 +166,14 @@ ZTEST(fail_in_setup, test_should_never_run)
+void main(void)
+{
+	/* run the tests */
+	ztest_main();
```
We should have a `ZTEST` here to assert that this main is the one that's running the tests.

yperess replied:
Agreed, this needs to assert something to count as a test.

yperess reviewed (changes requested):
This PR seems to work, but I'm not sold on why it's needed. Is the idea that you already have code with `main` that you cannot change, and you want to add tests to it? I think a sample or better documentation for why this is needed would go a long way.

fouge commented:
@yperess
I want to leverage ZTest in my main application without having to create new test applications so that tests are performed under my application context.
I could modify my code to not link my main function when ZTEST is enabled and use `test_main` instead but:
- I don't want the report to be printed only when returning from test_main(), but right after the tests are run (asap)
- I don't want to stop my application after tests are run

I could certainly find solutions to bypass the current behavior and make it work for my use case but I thought this was probably easier to remove definition of `main()` from ztest and make calling the tests easily accessible.

Let me know what you think

yperess commented:
> @yperess I want to leverage ZTest in my main application without having to create new test applications so that tests are performed under my application context. I could modify my code to not link my main function when ZTEST is enabled and use `test_main` instead but:
> 
> * I don't want the report to be printed only when returning from test_main(), but right after the tests are run (asap)

`test_main()` should be running the tests, is the issue that you're trying to merge it with your application logic so this function has extra latency (or doesn't return?)

> * I don't want to stop my application after tests are run
> 
> I could certainly find solutions to bypass the current behavior and make it work for my use case but I thought this was probably easier to remove definition of `main()` from ztest and make calling the tests easily accessible.
> 
> Let me know what you think

I have 2 concerns currently that I'm trying to work around:
1. We do have initialization code in ztest's `main()` function, so as that changes you (or any other downstream user) would have to keep up with the changes.
2. ztest tightly binds to twister so similar to issue #1, you might have to keep up with changes to other parts of ztest (not just the `main()` function).

Would you be able to rename your application's main function instead based `CONFIG_ZTEST`? So what I'm suggesting is:

### Rename your main entry point
```
#ifdef CONFIG_ZTEST
void app_main(void *arg0, void *arg1, void *arg2)
#else
void main(void)
#endif
```

### Add a new thread that's only linked in when testing
```
/*
 * Include the following in your CMakeLists.txt:
 *     target_add_sources_ifdef(CONFIG_ZTEST app test_main.c)
 */

extern void app_main(void *, void *, void *);
K_THREAD_DEFINE(app_main_tid, CONFIG_MAIN_STACK_SIZE,
    app_main, NULL, NULL, NULL,
    CONFIG_MAIN_THREAD_PRIORITY, 0, 0);

void test_main()
{
  /*
   * Use a condvar here to wait for the tests to all run,
   * you'll need to trigger that from app_main or where
   * ever in your application you call:
   *     ztest_run_registered_test_suites()
   */
}
```

It's not as pretty as just replacing the `main()` function, but it should give you more flexibility without having to debug strange errors if we ever make changes to ztest. Alternatively (not trying to dissuade you from using ztest) if you don't want to use twister, you could link in gTest instead which would give you the additional flexibility you want, in theory it's possible to modify gTest to work with twister but I haven't had time to look into it.

fouge commented:
@yperess I understand your concerns
I am closing so that we don't increase complexity over time 😉
Thanks!
